### ISSUE TITLE
Pasting text from a Qt application to Firefox sometimes fails

### ISSUE THREAD

**Issue by tgn-ff** [fcitx/fcitx5#702]
STR:
1. Open a Firefox window.
2. Copy some text from a Qt application. I used KDE's Kate.
3. Paste the text in Firefox's urlbar.
4. Copy some other text from the Qt application.
5. Open a new tab in Firefox and try to paste it in the urlbar.

Results:

The text needs a lot of tries (`Ctrl + V`) before it is, actually, pasted. fitx5 needs to be running, but not necessarily be in use, ie. it happens when English is the selected language (with the IME off).

One workaround is to set `Ctrl + V` (same as system's clipboard) as the "Paste Primary" shortcut in the Clipboard Module. I'm not sure if this causes some other input issues I have, though.

-----

I ran a bisection and ended up at the following commit as the culprit:

```c7725763d1b608161233ef62396a073a6004c27d is the first bad commit
commit c7725763d1b608161233ef62396a073a6004c27d
Author: Weng Xuetian <wengxt@gmail.com>
Date:   Sat Nov 12 09:58:04 2022 -0800

    Seems firefox doesn't like the batch mode call sequence.
    
    ProcessKeyEventBatch is introduced to mainly workaround Gtk4 limitation.
    While it does helps and ideally it should work on gtk2, and gtk3,
    certain app (namely firefox) doesn't like it. So lets just revert gtk2/3
    back to the old behavior, since they have been working well and no one
    complain about it. hangul case is tested, gtk3 works well and
    fcitx/fcitx5#445 does not happen.
    
    Fix fcitx/fcitx5-gtk#22

 fcitx-gclient/fcitxgclient.c | 45 ++++++++++++++++++++++++++++++++++++--------
 fcitx-gclient/fcitxgclient.h |  2 ++
 gtk2/fcitximcontext.cpp      |  1 +
 gtk3/fcitximcontext.cpp      |  1 +
 4 files changed, 41 insertions(+), 8 deletions(-)```

**Comment by wengxt** [comment]
Can you try with the current stable version of fcitx5-gtk and fcitx5-git to see if your issue is resolved?

**Comment by wengxt** [comment]
The observation from the log is that almost all key press event go through the key filtered code path, which is not my intention. The reason I think is because of some recent changes keyboard engine that always send preedit update.

In most case, preedit is simply empty and shouldn't trigger the re-forward code path. The changes above try to check if the pending event satisfies the condition and hopefully move keys like "ctrl + v" to another code path that doesn't trigger the issue.

**Comment by tgn-ff** [comment]
Yes, this fixes the pasting issue, but it somewhat re-introduces the previous one. Specifically, what's described in https://github.com/fcitx/fcitx5-gtk/issues/22#issuecomment-1312527298. With a couple of differences. 1) Pressing Enter won't submit the query, but ANY key (except for Backspace) pressed afterwards, will. Eg. With hangul, you can type `fk ENTER` and this will not submit the query, but if you press eg. `Ctrl`, it will. 2) This seems to fail only once. Every other query, even in new windows, will succeed until a browser restart.

**Comment by wengxt** [comment]
@tgn-ff sorry I made a small mistake in the previous commit, can you check it again with https://github.com/fcitx/fcitx5/commit/13f7d4c7fbd0056438efbbf91e740525f65f1337 ? (just rebuild fcitx5-git should be fine)

**Comment by tgn-ff** [comment]
Yes, both issues seem to be resolved now. However, there is a slight, but noticeable (>~150ms), delay when submitting queries through the urlbar. You can, actually, see the two steps that take place: committing the syllable then submitting the query.

**Comment by wengxt** [comment]
@tgn-ff  can you do a screen record demostrate the issue you describe in the last comment? I'm not sure what do you mean.

**Comment by wengxt** [comment]
Also, urlbar delay you described, is it new issue? or an existing issue?

**Comment by wengxt** [comment]
I feel it's all expected from what I saw in firefox.
1. commit text 
2. triggers omni urlbar search in firefox, which could be slow, and seems it wants combobox to be shown first.
3. firefox handles enter.

Firefox seems doesn't implement a "blindly launch the search in the background without updating UI". I believe it's firefox's job to optimize their workflow.

Chrome bar seems does much better on this. After commit text + enter it doesn't need to show the combobox. (feels like the combobox is implemented asynchronously)

There's an about:config option in firefox, which allows to search with uncommitted preedit: keepPanelOpenDuringImeComposition. It caused issue to input method in the past https://bugzilla.mozilla.org/show_bug.cgi?id=1693726, but seems to be fixed. You can give it a try though I don't see it improve anything on the speed on my side, the combobox doesn't show and hide though if you enables it.

**Comment by tgn-ff** [comment]
You may be right. In any case, the delay is not new and it affects other IMEs as well; I just noticed it now due to the extra attention I paid. I'll see if it's worth to file a bug against Firefox at some point. Thank you for the patch.
